File: java/src/main/java/io/binac/leetcode/SortColors.java
package io.binac.leetcode;

import java.util.Arrays;

/**
 * <p>Given an array <code>nums</code> with <code>n</code> objects colored red, white, or blue, sort them <strong><a href="https://en.wikipedia.org/wiki/In-place_algorithm" target="_blank">in-place</a> </strong>so that objects of the same color are adjacent, with the colors in the order red, white, and blue.</p>
 *
 * <p>We will use the integers <code>0</code>, <code>1</code>, and <code>2</code> to represent the color red, white, and blue, respectively.</p>
 *
 * <p>You must solve this problem without using the library's sort function.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> nums = [2,0,2,1,1,0]
 * <strong>Output:</strong> [0,0,1,1,2,2]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> nums = [2,0,1]
 * <strong>Output:</strong> [0,1,2]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>n == nums.length</code></li>
 * 	<li><code>1 &lt;= n &lt;= 300</code></li>
 * 	<li><code>nums[i]</code> is either <code>0</code>, <code>1</code>, or <code>2</code>.</li>
 * </ul>
 *
 * <p>&nbsp;</p>
 * <p><strong>Follow up:</strong>&nbsp;Could you come up with a one-pass algorithm using only&nbsp;constant extra space?</p>
 */
public class SortColors {
    public static class Solution1 {
        public void sortColors(int[] nums) {
            int count0 = 0, count1 = 0;
            for (int num : nums) {
                if (num == 0)
                    ++count0;
                else if (num == 1)
                    ++count1;
            }
            Arrays.fill(nums, 0, count0, 0);
            Arrays.fill(nums, count0, count0 + count1, 1);
            Arrays.fill(nums, count0 + count1, nums.length, 2);
        }
    }

    public static class Solution2 {
        private void swap(int[] x, int a, int b) {
            int t = x[a];
            x[a] = x[b];
            x[b] = t;
        }

        public void sortColors(int[] nums) {
            int last0 = -1, first2 = nums.length;
            for (int i = 0; i < first2; ) {
                if (nums[i] == 0)
                    swap(nums, ++last0, i++);
                else if (nums[i] == 2)
                    swap(nums, i, --first2);
                else
                    ++i;
            }
        }
    }
}

File: python3/leetcodepy/sort_colors.py
"""
Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.

You must solve this problem without using the library's sort function.



Example 1:

Input: nums = [2,0,2,1,1,0]
Output: [0,0,1,1,2,2]

Example 2:

Input: nums = [2,0,1]
Output: [0,1,2]


Constraints:

n == nums.length
1 <= n <= 300
nums[i] is either 0, 1, or 2.


Follow up: Could you come up with a one-pass algorithm using only constant extra space?
"""
from typing import List


class Solution1:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        count0, count1 = 0, 0
        for n in nums:
            if n == 0:
                count0 += 1
            elif n == 1:
                count1 += 1
        for i in range(0, count0):
            nums[i] = 0
        for i in range(count0, count0 + count1):
            nums[i] = 1
        for i in range(count0 + count1, len(nums)):
            nums[i] = 2


class Solution2:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        last0, i, first2 = -1, 0, len(nums)
        while i < first2:
            if nums[i] == 0:
                last0 += 1
                nums[last0], nums[i] = nums[i], nums[last0]
                i += 1
            elif nums[i] == 2:
                first2 -= 1
                nums[i], nums[first2] = nums[first2], nums[i]
            else:
                i += 1
